include/verilated_random.h

```diff
@@ -476,18 +482,20 @@ class VlRandomizer final {
                       std::vector<size_t> idxWidths) {
         std::ostringstream oss;
         for (size_t i = 0; i < indices.size(); ++i) {
-            oss << std::hex << static_cast<int>(indices[i]);
+            oss << std::hex << std::setw(int(idxWidths[i] / 4)) << std::setfill('0')
+                << static_cast<int>(indices[i]);
             if (i < indices.size() - 1) oss << ".";
         }
-        write_var(var, 1ULL, (name + "." + oss.str()).c_str(), 1ULL);
+        write_var(var, 1ULL,
+                  oss.str().length() > 0 ? (name + "." + oss.str()).c_str() : name.c_str(), 1ULL);
     }
 
     // Recursively process VlUnpacked of structs
     template <typename T, std::size_t N_Depth>
     void record_struct_arr(VlUnpacked<T, N_Depth>& var, const std::string name, int dimension,
                            std::vector<IData> indices, std::vector<size_t> idxWidths) {
         if (dimension > 0 && N_Depth != 0) {
-            idxWidths.push_back(32);
+            idxWidths.push_back((1 << (int)ceil(log2(int(log2(N_Depth)) + 1))));
```

Comment:
Please use the existing VL_ functions to get to the size you need.  Floating point verisons are slow, and break with larger numbers.